src/main/java/net/snowflake/client/jdbc/SnowflakeUtil.java

```diff
@@ -263,42 +311,65 @@ public static SnowflakeColumnMetadata extractColumnMetadata(
               colType = Types.BINARY;
           }
         }
+        columnTypeInfo = new ColumnTypeInfo(colType, extColTypeName, baseType);
         break;
 
       default:
         throw new SnowflakeSQLLoggedException(
-            session,
+            new SFSession(),
             ErrorCode.INTERNAL_ERROR.getMessageCode(),
             SqlState.INTERNAL_ERROR,
             "Unknown column type: " + internalColTypeName);
     }
 
-    JsonNode extColTypeNameNode = colNode.path("extTypeName");
-    if (!extColTypeNameNode.isMissingNode()
-        && !Strings.isNullOrEmpty(extColTypeNameNode.asText())) {
-      extColTypeName = extColTypeNameNode.asText();
-    }
-
-    String colSrcDatabase = colNode.path("database").asText();
-    String colSrcSchema = colNode.path("schema").asText();
-    String colSrcTable = colNode.path("table").asText();
+    return columnTypeInfo;
+  }
 
-    boolean isAutoIncrement = colNode.path("isAutoIncrement").asBoolean();
+  private static String defaultIfNull(String extColTypeName, String defaultValue) {
+    return Optional.ofNullable(extColTypeName).orElse(defaultValue);
+  }
 
-    return new SnowflakeColumnMetadata(
-        colName,
-        colType,
-        nullable,
-        length,
-        precision,
-        scale,
-        extColTypeName,
-        fixed,
-        baseType,
-        colSrcDatabase,
-        colSrcSchema,
-        colSrcTable,
-        isAutoIncrement);
+  static FieldMetadata[] createFieldsMetadata(ArrayNode fieldsJson, int fixedColType)
+      throws SnowflakeSQLLoggedException {
+    //    TODO: verify that jsonFileds is not empty array
+    FieldMetadata[] fields = new FieldMetadata[fieldsJson.size()];
+    int fieldCounter = 0;
+    for (JsonNode node : fieldsJson) {
+      String colName = node.path("name").asText();
+      int scale = node.path("scale").asInt();
+      int precision = node.path("precision").asInt();
+      String internalColTypeName = node.path("type").asText();
+      boolean nullable = node.path("nullable").asBoolean();
+      int length = node.path("length").asInt();
+      boolean fixed = node.path("fixed").asBoolean();
+      FieldMetadata[] internalFields = new FieldMetadata[0];
```

Comment:
let's extract method and create variable once

Reply:
Fixed